core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -374,37 +372,33 @@ class TransactionStateManager(brokerId: Int,
       }
 
       if (responseError == Errors.NONE) {
-        def completeStateTransition(metadata: TransactionMetadata, newState: TransactionState): Boolean = {
-          // there is no transition in this case
-          if (metadata.state == Empty && newState == Empty)
-            true
-          else
-            metadata.completeTransitionTo(txnMetadata.state)
-        }
         // now try to update the cache: we need to update the status in-place instead of
         // overwriting the whole object to ensure synchronization
-          getTransactionState(transactionalId) match {
-            case Some(metadata) =>
-              metadata synchronized {
-                if (metadata.pid == txnMetadata.pid &&
-                  metadata.producerEpoch == txnMetadata.producerEpoch &&
-                  metadata.txnTimeoutMs == txnMetadata.txnTimeoutMs &&
-                  completeStateTransition(metadata, txnMetadata.state)) {
-                  // only topic-partition lists could possibly change (state should have transited in the above condition)
-                  metadata.addPartitions(txnMetadata.topicPartitions.toSet)
-                } else {
-                  throw new IllegalStateException(s"Completing transaction state transition to $txnMetadata while its current state is $metadata.")
-                }
+        getTransactionState(transactionalId) match {
+          case Some(epochAndMetadata) =>
+            epochAndMetadata synchronized {
+              val metadata = epochAndMetadata.transactionMetadata
+
+              if (epochAndMetadata.coordinatorEpoch == coordinatorEpoch && metadata.completeTransitionTo(txnMetadata)) {
```

Comment:
Not sure I understand your question? `coordinatorEpoch` is the value passed from the caller, which is the epoch of the cached metadata object before append is called, and here we are checking if it is still the same epoch by reading from the cache again. Plus inside `completeTransitionTo` we also check that other fields inside metadata are expected.